client/src/file/locked_keyring.rs

```diff
@@ -0,0 +1,136 @@
+                    n_broken_items += 1;
+                } else {
+                    n_valid_items += 1;
+                }
+            }
+
+            drop(inner_keyring);
+
+            if n_valid_items == 0 && n_broken_items != 0 {
+                #[cfg(feature = "tracing")]
+                tracing::error!("Keyring cannot be decrypted. Invalid secret.");
+                return Err(Error::IncorrectSecret);
+            } else if n_broken_items > n_valid_items {
+                #[cfg(feature = "tracing")]
+                {
+                    tracing::warn!(
+                        "The file contains {n_broken_items} broken items and {n_valid_items} valid ones."
+                    );
+                    tracing::info!(
+                        "Please switch to `Keyring::load_unchecked` to load the keyring without the secret validation.
+                        `Keyring::delete_broken_items` can be used to remove them or alternatively with `oo7-cli --repair`."
+                    );
+                }
+                return Err(Error::IncorrectSecret);
+            }
+            Some(Arc::new(key))
+        } else {
+            None
+        };
+
+        Ok(Keyring {
+            keyring: self.keyring,
+            path: self.path,
+            mtime: self.mtime,
+            key: Mutex::new(key),
+            secret: Mutex::new(Arc::new(secret)),
+        })
+    }
+
+    pub async fn load(path: PathBuf) -> Result<Self, Error> {
+        let (mtime, keyring) = match fs::File::open(&path).await {
+            Err(err) if err.kind() == io::ErrorKind::NotFound => {
+                #[cfg(feature = "tracing")]
+                tracing::debug!("Keyring file not found, creating a new one");
+                (None, api::Keyring::new())
+            }
+            Err(err) => return Err(err.into()),
+            Ok(mut file) => {
+                #[cfg(feature = "tracing")]
+                tracing::debug!("Keyring file found, loading it content");
+                let mtime = file.metadata().await?.modified().ok();
+
+                let mut content = Vec::new();
+                file.read_to_end(&mut content).await?;
+
+                let keyring = api::Keyring::try_from(content.as_slice())?;
+
+                (mtime, keyring)
+            }
+        };
+
+        Ok(Self {
+            keyring: Arc::new(RwLock::new(keyring)),
+            path: Some(path),
+            mtime: Mutex::new(mtime),
+        })
+    }
+
+    pub async fn open(name: &str) -> Result<Self, Error> {
+        let v1_path = api::Keyring::path(name, api::MAJOR_VERSION)?;
+        if v1_path.exists() {
+            return Self::load(v1_path).await;
+        }
+
+        let v0_path = api::Keyring::path(name, api::LEGACY_MAJOR_VERSION)?;
+        if v0_path.exists() {
+            // Migrating a locked keyring is more involved without having the
+            // secret.
+            todo!();
+        };
```

Comment:
There is a fundamental difference between v0 & v1. v0 encrypts the whole keyring, v1 encrypts individual items. So we cannot really achieve this.

I am leaning towards not providing the open api for locked keyring and only do that for cases where you also provide a secret.

Reply:
The migration step should be moved for a time after we have access to the password, or we somehow prompt the user for password for the migration+unlock.

Reply:
I mean, this is for backward compatibility purposes only, we can take care of that on the server side.

Reply:
We can for now provide a method that also does the migration from v0 to v1 that takes the Secret. The server side can use that instead of the common use case, the file backend backed by the portal.